src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Legacy/LegacySyntaxNodeExtensions.NodeStack.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+
+namespace Microsoft.AspNetCore.Razor.Language.Legacy;
+
+internal static partial class LegacySyntaxNodeExtensions
+{
+    private struct NodeStack : IDisposable
```

Comment:
What's the advantage of introducing this type over just using pooled `ArrayBuilder`s?

Reply:
We don't have `ArrayBuilder`s in Razor.

Reply:
If we referenced `System.Memory`, I would at least use `ArrayPool<SyntaxNode>`.

Reply:
I've switched over to using an `ObjectPool<SyntaxNode[]>`. I'm happy to move over to an `ObjectPool<Immutable<SyntaxNode>.Builder>` once your PR is merged.

Reply:
Thanks :). The threadstatic made me nervous.